## Analysing atp tennis matches with pandas<br><br> 

The dataset we will manipulate with consists of men's tennis matches at Wimbledon.<br><br>Many different approaches are possible so what I am trying to do here is to draw the attention to points that might not seem to be obvious at the first glance.<br>The data I am using can be downloaded in .csv format from the following link.<br>

<br><br>https://github.com/solajozsef/hello-world/blob/main/AtpWimbledonMatches.csv<br><br>If you need more data it is available at<br><br>http://tennis-data.co.uk<br><br>and there is a 'notes.txt' page explaining the abbreviations on <br><br>http://tennis-data.co.uk/notes.txt<br>

### Importing some libraries

In [1]:
import numpy as np
import pandas as pd
import math
import random
import glob

<br>This code provides multiple outputs from one code cell.

In [2]:
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"

In [3]:
pd.set_option("display.max_rows", 3000, "display.max_columns", 50)
pd.get_option("display.max_rows"),
pd.get_option("display.max_columns")

(3000,)

50

<br>The dataset we are using can be downloaded from the following github link.<br><br>

In [6]:
url = 'https://raw.githubusercontent.com/paplak/hello-world/main/AtpWimbledonMatches.csv'

<br>Because it is a .csv file we have to read it in pandas dataframe.<br><br>

In [8]:
df = pd.read_csv(url)

<br>Some basic info about df columns.<br><br>

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1905 entries, 0 to 1904
Data columns (total 17 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Location    1905 non-null   object 
 1   Tournament  1905 non-null   object 
 2   Date        1905 non-null   object 
 3   Winner      1905 non-null   object 
 4   Loser       1905 non-null   object 
 5   WRank       1905 non-null   int64  
 6   LRank       1905 non-null   int64  
 7   W1          1898 non-null   float64
 8   L1          1898 non-null   float64
 9   W2          1885 non-null   float64
 10  L2          1886 non-null   float64
 11  W3          1858 non-null   float64
 12  L3          1858 non-null   float64
 13  Wsets       1899 non-null   float64
 14  Lsets       1899 non-null   float64
 15  B365W       1897 non-null   float64
 16  B365L       1899 non-null   float64
dtypes: float64(10), int64(2), object(5)
memory usage: 253.1+ KB


In [7]:
df.shape

(1905, 17)

<br>As we see there are 1905 rows and 17 columns in the dataframe we are going to play with.<br><br>

In [8]:
df.describe()

WRank        LRank           W1           L1           W2  \
count  1905.000000  1905.000000  1898.000000  1898.000000  1885.000000   
mean     49.756955    88.554331     5.800843     4.278714     5.855703   
std      70.233601    95.966322     1.196269     1.815351     1.184819   
min       1.000000     1.000000     0.000000     0.000000     0.000000   
25%      11.000000    36.000000     6.000000     3.000000     6.000000   
50%      30.000000    70.000000     6.000000     4.000000     6.000000   
75%      68.000000   109.000000     6.000000     6.000000     6.000000   
max    1065.000000  1085.000000     7.000000     7.000000     7.000000   

                L2           W3           L3        Wsets        Lsets  \
count  1886.000000  1858.000000  1858.000000  1899.000000  1899.000000   
mean      4.229056     5.824004     4.043595     2.938915     0.678252   
std       1.792565     1.142381     1.835665     0.344239     0.772421   
min       0.000000     1.000000     0.000000     0.000000     0.000000   
25%       3.000000     6.000000     3.000000     3.000000     0.000000   
50%       4.000000     6.000000     4.000000     3.000000     0.000000   
75%       6.000000     6.000000     6.000000     3.000000     1.000000   
max       7.000000     7.000000     7.000000     3.000000     2.000000   

             B365W        B365L  
count  1897.000000  1899.000000  
mean      1.777132     5.196707  
std       1.696712     5.187248  
min       1.000000     1.002000  
25%       1.110000     2.000000  
50%       1.300000     3.400000  
75%       1.720000     6.500000  
max      29.000000    34.000000

<br>We can check the first 5 rows of df.<br><br>

In [9]:
df.head()

Location Tournament      Date         Winner             Loser  WRank  \
0   London  Wimbledon  29/06/15      Broady L.      Matosevic M.    182   
1   London  Wimbledon  29/06/15       Cilic M.         Moriya H.      9   
2   London  Wimbledon  29/06/15    Berankis R.  Haider-Maurer A.     90   
3   London  Wimbledon  29/06/15       Thiem D.           Sela D.     30   
4   London  Wimbledon  29/06/15  Granollers M.     Tipsarevic J.     72   

   LRank   W1   L1   W2   L2   W3   L3  Wsets  Lsets  B365W  B365L  
0    138  5.0  7.0  4.0  6.0  6.0  3.0    3.0    2.0   2.62   1.44  
1    174  6.0  3.0  6.0  2.0  7.0  6.0    3.0    0.0   1.02  17.00  
2     57  6.0  2.0  5.0  2.0  NaN  NaN    1.0    0.0   1.50   2.50  
3     85  2.0  6.0  6.0  3.0  6.0  4.0    3.0    1.0   1.53   2.37  
4    486  6.0  3.0  6.0  4.0  6.0  2.0    3.0    0.0   1.57   2.25

<br>We can collect df's column names with<br><br>

In [10]:
df.columns

Index(['Location', 'Tournament', 'Date', 'Winner', 'Loser', 'WRank', 'LRank',
       'W1', 'L1', 'W2', 'L2', 'W3', 'L3', 'Wsets', 'Lsets', 'B365W', 'B365L'],
      dtype='object')

<br>I like changing capital letters in columns' names, makes life easier.<br><br>

In [11]:
df.columns = df.columns.str.lower()

In [12]:
df.columns

Index(['location', 'tournament', 'date', 'winner', 'loser', 'wrank', 'lrank',
       'w1', 'l1', 'w2', 'l2', 'w3', 'l3', 'wsets', 'lsets', 'b365w', 'b365l'],
      dtype='object')

<br>So let's get to it.<br>Let's see how many favorite players with odds lower than say 1.30 won the match and how many lost. The winners odds are in the 'b365w' column.<br>To get a clear view we need to know how many players with odds lower than 1.3 are in the winners' column(b365w) and how many are in the losers' column(b365l).<br><br>

In [13]:
df.query('(b365w <= 1.3) or (b365l <= 1.3) ').b365w.count()

1104

In [14]:
df.query('(b365w <= 1.3)').b365w.count()

961

<br>When the favorite loses his odds appear in the 'b365l' column.<br><br>

In [15]:
df.query('(b365l <= 1.3)').b365l.count()

143

<br>What we see here is that out of 1104 matches the favorites won 961 matches and lost 143 ones.<br><br>

<br>How many winners' ranks are above say 900 in world ranking.<br><br>

In [16]:
df[df['wrank'] > 900].count().b365w

2

<br>and who are those...<br><br>

In [17]:
df[df['wrank'] > 900]

location tournament      date     winner        loser  wrank  lrank   w1  \
1320   London  Wimbledon  26/06/07  Kiefer N.  Volandri F.   1065     27  6.0   
1354   London  Wimbledon  28/06/07  Kiefer N.   Santoro F.   1065     70  6.0   

       l1   w2   l2   w3   l3  wsets  lsets  b365w  b365l  
1320  3.0  7.0  6.0  6.0  1.0    3.0    0.0   1.16   4.50  
1354  4.0  6.0  3.0  6.0  4.0    3.0    0.0   2.37   1.53

<br>Say I would like to see the favorite losers where the favorite odds were between 1.4 and 1.5<br><br>

In [18]:
df.query('(b365l >= 1.4) and (b365l < 1.5)').count().b365w

51

<br>... and the total number of games within these favorite odds<br><br>

In [19]:
df.query('((b365w >= 1.4) and (b365w < 1.5)) or ((b365l >= 1.4) and (b365l < 1.5))').count().b365w

150

<br>So altogether there were 150 games within these favorite odds and the favorite won 51 matches.<br><br>

<br>Underdogs with odds higher than 4.0 winning the match. <br><br>

In [20]:
df.query('b365w >= 4').count().b365w

109

<br>And underdogs with odds higher than 4.0 losing the match. <br><br>

In [21]:
df.query('b365l >= 4').count().b365w

845

<br>Let's see how tennis aces perfom against each other. The following cell filters those games where players' world rankings were between 1 and 10 and we printed the relevant cells.<br><br>wrank: winner's rank<br>lrank: loser's rank<br>b365w: winner's odds<br>b365l: loser's odds<br><br>The first column is the row number column in our table.<br>

In [22]:
df.query('((wrank >= 1) and (wrank <= 10 )) and ((lrank >= 1) and (lrank <= 10 ) and (b365w < b365l))')[['wrank', 'lrank', 'b365w', 'b365l']],

(      wrank  lrank  b365w  b365l
 123       1      9   1.08   7.50
 126       1      2   1.83   2.00
 247       1     10   1.05   9.00
 253       1      2   1.16   5.00
 380       2      6   1.11   6.50
 505       1      2   1.11   6.50
 507       1      4   1.14   5.50
 630       1      6   1.66   2.20
 631       4     10   1.22   4.33
 633       1      4   1.57   2.37
 757       3      7   1.11   6.50
 761       1      3   1.55   2.60
 884       5      7   1.20   4.50
 888       5      6   1.22   4.50
 1011      1      9   1.14   5.50
 1014      1      4   1.50   2.62
 1136      1      6   1.10   7.00
 1140      1      8   1.16   5.00
 1268      2      9   1.10   7.00
 1269      2      7   1.25   4.33
 1382      1     10   1.01  15.00
 1395      2      5   1.25   4.00
 1396      1      2   1.14   5.50
 1519      4      3   1.28   3.75
 1522      4      9   1.40   3.00
 1523      2      4   1.61   2.40
 1646      1      4   1.36   3.20
 1900      4      5   1.44   2.75
 1903      4  

In [23]:
df.query('((wrank >= 1) and (wrank <= 10 )) and ((lrank >= 1) and (lrank <= 10 ) and (b365l < b365w))')[['wrank', 'lrank', 'b365w', 'b365l']],

(      wrank  lrank  b365w  b365l
 379       6      3   3.50   1.30
 760       3      2   2.10   1.72
 1005      9      7   2.87   1.40
 1015      2      1   2.30   1.66
 1137      8      4   2.10   1.72
 1142      2      1   2.50   1.57
 1267      7      3   2.62   1.50
 1645      8      2  10.00   1.07
 1648      8     10   2.00   1.80
 1777      2      1   2.20   1.66
 1897      5      9   2.20   1.66
 1902      3      1   2.87   1.44,)

<br>The first table shows the matches where the favorite won and the second where the favorite lost.<br>

Let's focus on very strong favorites whose booky odds are below 1.10 and see how they perform.<br>First we check how many matches they won.<br><br>

In [24]:
df.query('b365w <= 1.1').count().b365w

458

<br>Now how many matches they lost<br><br>

In [25]:
df.query('b365l <= 1.1').count().b365w

24

<br>The figures are quite impressive. We can say that strong favorites with odds lower than 1.10 win the majority of the games. <br><br>

In [26]:
df.query('b365l <= 1.1').head()

location tournament      date            winner          loser  wrank  \
29    London  Wimbledon  29/06/15          Ilhan M.    Janowicz J.     82   
46    London  Wimbledon  30/06/15  Ramos-Vinolas A.     Istomin D.     65   
223   London  Wimbledon  30/06/06       Verdasco F.  Nalbandian D.     30   
317   London  Wimbledon  24/06/09  Gimeno-Traver D.        Dent T.     98   
464   London  Wimbledon  23/06/05       Tursunov D.      Henman T.    152   

     lrank   w1   l1   w2   l2   w3   l3  wsets  lsets  b365w  b365l  
29      47  7.0  6.0  6.0  4.0  6.0  7.0    3.0    1.0    7.0   1.10  
46      62  6.0  2.0  6.0  2.0  3.0  2.0    2.0    0.0    7.0   1.10  
223      3  7.0  6.0  7.0  6.0  6.0  2.0    3.0    0.0    7.0   1.08  
317    266  7.0  5.0  7.0  6.0  4.0  6.0    3.0    2.0    7.0   1.08  
464      9  3.0  6.0  6.0  2.0  3.0  6.0    3.0    2.0    7.0   1.08

<br>We can search for players by name like so<br><br>

In [27]:
name = 'Nadal'
wn = df.winner.str.contains(name)
ls = df.loser.str.contains(name)

<br>Above we created two variables 'wn' and 'ln'. 'wn' stands for winner's name and 'ln' for loser's name.<br>If we want to see how many times this player won or lost, the code is below.<br>Note: if we want to refer to a variable within a query we use the '@' character<br><br>

In [28]:
print('win: ', df.query('@wn == True').shape[0],  'times')
print('lose: ', df.query('@ls == True').shape[0], 'times')

win:  51 times
lose:  11 times


<br>The same process with Thiem.<br><br>

In [29]:
name = 'Thiem'
wn = df.winner.str.contains(name)
ls = df.loser.str.contains(name)

In [30]:
print('win: ', df.query('@wn == True').shape[0],  'times')
print('lose: ', df.query('@ls == True').shape[0], 'times')

win:  5 times
lose:  6 times


<br>We can even write a little function for this operation.<br><br>

In [31]:
def win_lose(name):
    wn = df.winner.str.contains(name)
    ls = df.loser.str.contains(name)
    
    if df.winner.str.contains(name).any():
      #  var = df.query('@mm == True').shape[0]
        print('win:  ', df.query('@wn == True').shape[0],  'times')
        
    if df.loser.str.contains(name).any():
       # var = df.query('@mm == True').shape[0]
        print('lose: ', df.query('@ls == True').shape[0],  'times')

<br>and then we only have to modify the 'name' variable.<br><br>

In [34]:
name = 'Cilic'
win_lose(name)

win:   29 times
lose:  13 times


In [35]:
name = 'Berankis'
win_lose(name)

win:   3 times
lose:  7 times


In [37]:
name = 'Djoko'
win_lose(name)

win:   72 times
lose:  10 times


<br>If you want to know who has defeated Djokovic one possible way is below:<br><br>

In [40]:
df[df.loser.str.contains('Djokovic') == True][['winner', 'loser', 'b365w', 'b365l']]

winner        loser  b365w  b365l
243      Ancic M.  Djokovic N.   1.36   3.00
375       Haas T.  Djokovic N.   3.20   1.36
486   Grosjean S.  Djokovic N.   1.14   5.00
632    Berdych T.  Djokovic N.   2.00   1.80
885    Berdych T.  Djokovic N.   4.00   1.25
1142    Murray A.  Djokovic N.   2.50   1.57
1243   Querrey S.  Djokovic N.  26.00   1.01
1395     Nadal R.  Djokovic N.   1.25   4.00
1718     Safin M.  Djokovic N.   9.00   1.05
1902   Federer R.  Djokovic N.   2.87   1.44

<br>Modify the code blocks as you like!